clients/src/main/java/org/apache/kafka/clients/consumer/internals/ShareConsumeRequestManager.java

```diff
@@ -815,21 +816,26 @@ private void handleShareFetchSuccess(Node fetchTarget,
                     }
                 }
 
-                ShareCompletedFetch completedFetch = new ShareCompletedFetch(
+                completedFetches.add(
+                    new ShareCompletedFetch(
                         logContext,
                         BufferSupplier.create(),
                         fetchTarget.id(),
                         tip,
                         partitionData,
                         shareFetchMetricsAggregator,
-                        requestVersion);
-                shareFetchBuffer.add(completedFetch);
+                        requestVersion)
+                );
 
                 if (!partitionData.acquiredRecords().isEmpty()) {
                     fetchMoreRecords = false;
                 }
             }
 
+            if (!completedFetches.isEmpty()) {
+                shareFetchBuffer.add(completedFetches);
```

Comment:
Is this primarily a performance optimization for handling many partitions? If so, should we reduce the `ShareAcknowledgementCommitCallbackEvent` by batching partitions ?

Reply:
It's more to do with correctness. I'm going back into the callback stuff shortly, so I'll bear in mind your comment.